File: sdks/java/src/main/java/com/blok/blok/node/TypedNode.java
package com.blok.blok.node;

import com.blok.blok.errors.BlokError;
import com.blok.blok.types.Context;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.util.List;
import java.util.Map;

/**
 * Typed node base (SPEC-B P4) — the Java equivalent of the TypeScript
 * {@code defineNode} / Python {@code @node} / Rust {@code TypedNode} / Go
 * {@code DefineNode}. Declare typed {@code I}/{@code O} types (e.g. records) and
 * implement {@link #run}; the SDK converts the raw config into the typed input
 * BEFORE running (a type mismatch → structured {@link BlokError}, HTTP 400),
 * and reflects both JSON Schemas for the catalog — instead of a raw
 * {@code Map<String, Object>}.
 *
 * <p>Register like any handler:
 * {@code registry.register("@acme/search", new SearchNode());}
 *
 * <pre>{@code
 * public record SearchInput(String query, int limit) {}
 * public record SearchOutput(List<String> results, int count) {}
 *
 * public final class SearchNode extends TypedNode<SearchInput, SearchOutput> {
 *     public String name() { return "@acme/search"; }
 *     public String description() { return "Full-text search"; }
 *     protected Class<SearchInput> inputClass() { return SearchInput.class; }
 *     protected Class<?> outputClass() { return SearchOutput.class; }
 *     protected SearchOutput run(Context ctx, SearchInput input) {
 *         var rows = Collections.nCopies(input.limit(), input.query());
 *         return new SearchOutput(rows, rows.size());
 *     }
 * }
 * }</pre>
 *
 * @param <I> the typed input
 * @param <O> the typed output
 */
public abstract class TypedNode<I, O> implements NodeHandler, NodeReflector {

    private static final Gson GSON = new Gson();

    /** The node's registered name (e.g. {@code "@acme/search"}). */
    public abstract String name();

    /** {@inheritDoc} Human-readable description; override to set. */
    @Override
    public String description() {
        return "";
    }

    /** The class of the typed input (e.g. {@code SearchInput.class}). */
    protected abstract Class<I> inputClass();

    /** The class of the typed output, for schema reflection; {@code null} to skip. */
    protected Class<?> outputClass() {
        return null;
    }

    /** Run the node with a VALIDATED, typed input. */
    protected abstract O run(Context ctx, I input) throws Exception;

    @Override
    public Object execute(Context ctx, Map<String, Object> config) throws Exception {
        I input;
        try {
            input = GSON.fromJson(GSON.toJsonTree(config), inputClass());
        } catch (RuntimeException e) {
            throw BlokError.validation()
                    .code("NODE_INPUT_VALIDATION")
                    .message("Input validation failed for node '" + name() + "': " + e.getMessage())
                    .httpStatus(400)
                    .node(name())
                    .build();
        }
        return run(ctx, input);
    }

    @Override
    public String inputSchemaJson() {
        return reflectSchema(inputClass());
    }

    @Override
    public String outputSchemaJson() {
        Class<?> out = outputClass();
        return out == null ? null : reflectSchema(out);
    }

    /** Build a minimal JSON Schema from the type's declared instance fields. */
    private static String reflectSchema(Class<?> type) {
        try {
            JsonObject schema = new JsonObject();
            schema.addProperty("type", "object");
            JsonObject properties = new JsonObject();
            JsonArray required = new JsonArray();
            for (Field field : type.getDeclaredFields()) {
                if (Modifier.isStatic(field.getModifiers())) {
                    continue;
                }
                JsonObject property = new JsonObject();
                property.addProperty("type", jsonType(field.getType()));
                properties.add(field.getName(), property);
                if (field.getType().isPrimitive()) {
                    required.add(field.getName());
                }
            }
            schema.add("properties", properties);
            if (required.size() > 0) {
                schema.add("required", required);
            }
            return GSON.toJson(schema);
        } catch (RuntimeException e) {
            return null;
        }
    }

    private static String jsonType(Class<?> type) {
        if (type == int.class || type == Integer.class || type == long.class || type == Long.class
                || type == short.class || type == Short.class) {
            return "integer";
        }
        if (type == double.class || type == Double.class || type == float.class || type == Float.class) {
            return "number";
        }
        if (type == boolean.class || type == Boolean.class) {
            return "boolean";
        }
        if (type == String.class) {
            return "string";
        }
        if (type.isArray() || List.class.isAssignableFrom(type)) {
            return "array";
        }
        return "object";
    }
}

File: sdks/php/src/Node/TypedNode.php
<?php

declare(strict_types=1);

namespace Blok\Blok\Node;

use Blok\Blok\Errors\BlokError;
use Blok\Blok\Types\Context;

/**
 * Typed node base (SPEC-B P4) — the PHP equivalent of the TypeScript
 * `defineNode` / Python `@node` / Rust `TypedNode`. Declare a typed input DTO
 * (a class with constructor-promoted typed properties) via {@see inputClass};
 * the SDK hydrates the raw config array into it (a type mismatch / missing
 * required field → structured {@see BlokError}, HTTP 400) BEFORE {@see run},
 * and reflects the input/output JSON Schemas — instead of a raw `array $config`.
 *
 * Register like any handler: `$registry->register('@acme/search', new SearchNode());`
 *
 * ```php
 * final class SearchInput {
 *     public function __construct(public string $query, public int $limit = 10) {}
 * }
 * final class SearchNode extends TypedNode {
 *     public function name(): string { return '@acme/search'; }
 *     public function description(): string { return 'Full-text search'; }
 *     protected function inputClass(): string { return SearchInput::class; }
 *     protected function run(Context $ctx, object $input): mixed {
 *         $rows = array_fill(0, $input->limit, $input->query);
 *         return ['results' => $rows, 'count' => count($rows)];
 *     }
 * }
 * ```
 */
abstract class TypedNode implements NodeHandler, NodeReflector
{
    /** The node's registered name (e.g. `"@acme/search"`). */
    abstract public function name(): string;

    /** Human-readable description, surfaced in the node catalog. */
    public function description(): string
    {
        return '';
    }

    /** @return class-string FQCN of the input DTO. */
    abstract protected function inputClass(): string;

    /** @return class-string|null FQCN of the output DTO (for schema reflection), or null. */
    protected function outputClass(): ?string
    {
        return null;
    }

    /**
     * Run the node with a VALIDATED, hydrated input DTO.
     *
     * @param object $input instance of {@see inputClass}
     * @return mixed the output (array/scalar/object) included in the result
     */
    abstract protected function run(Context $ctx, object $input): mixed;

    public function execute(Context $ctx, array $config): mixed
    {
        try {
            $input = $this->hydrate($this->inputClass(), $config);
        } catch (\Throwable $e) {
            throw BlokError::validation()
                ->code('NODE_INPUT_VALIDATION')
                ->message(sprintf("Input validation failed for node '%s': %s", $this->name(), $e->getMessage()))
                ->httpStatus(400)
                ->node($this->name())
                ->build();
        }

        return $this->run($ctx, $input);
    }

    /** Hydrate a config array into the typed DTO; the typed ctor enforces types. */
    private function hydrate(string $class, array $config): object
    {
        $ref = new \ReflectionClass($class);
        $ctor = $ref->getConstructor();
        if ($ctor === null) {
            return $ref->newInstance();
        }

        $args = [];
        foreach ($ctor->getParameters() as $param) {
            $name = $param->getName();
            if (array_key_exists($name, $config)) {
                $args[] = $config[$name];
            } elseif ($param->isDefaultValueAvailable()) {
                $args[] = $param->getDefaultValue();
            } elseif ($param->allowsNull()) {
                $args[] = null;
            } else {
                throw new \InvalidArgumentException("missing required field '{$name}'");
            }
        }

        // A typed constructor throws TypeError on an unconvertible value.
        return $ref->newInstanceArgs($args);
    }

    public function inputSchema(): ?array
    {
        return $this->reflectSchema($this->inputClass());
    }

    public function outputSchema(): ?array
    {
        $class = $this->outputClass();

        return $class === null ? null : $this->reflectSchema($class);
    }

    /** @return array<string, mixed> a minimal JSON Schema from the DTO's typed ctor params. */
    private function reflectSchema(string $class): array
    {
        $ref = new \ReflectionClass($class);
        $properties = [];
        $required = [];

        $ctor = $ref->getConstructor();
        if ($ctor !== null) {
            foreach ($ctor->getParameters() as $param) {
                $properties[$param->getName()] = $this->typeToSchema($param->getType());
                if (!$param->isOptional() && !$param->allowsNull()) {
                    $required[] = $param->getName();
                }
            }
        }

        $schema = ['type' => 'object', 'properties' => (object) $properties];
        if ($required !== []) {
            $schema['required'] = $required;
        }

        return $schema;
    }

    /** @return array<string, mixed> */
    private function typeToSchema(?\ReflectionType $type): array
    {
        if ($type instanceof \ReflectionNamedType) {
            return match ($type->getName()) {
                'int' => ['type' => 'integer'],
                'float' => ['type' => 'number'],
                'bool' => ['type' => 'boolean'],
                'string' => ['type' => 'string'],
                'array' => ['type' => 'array'],
                default => ['type' => 'object'],
            };
        }

        return [];
    }
}
